controllers/operatorpolicy_status.go

```diff
@@ -0,0 +1,429 @@
+						}
+					}
+
+					if prevObj.Compliant != updatedObj.Compliant {
+						relObjsChanged = true
+					} else if prevObj.Reason != updatedObj.Reason {
+						relObjsChanged = true
+					}
+				}
+			}
+
+			if !nameFound {
+				relObjsChanged = true
+			}
+		}
+	} else {
+		relObjsChanged = true
+	}
+
+	if relObjsChanged {
+		// start with the related objects which do not match the currently considered kind
+		newRelObjs := make([]policyv1.RelatedObject, 0)
+
+		for idx, relObj := range policy.Status.RelatedObjects {
+			if _, matchedIdx := prevRelObjs[idx]; !matchedIdx {
+				newRelObjs = append(newRelObjs, relObj)
+			}
+		}
+
+		// add the new related objects
+		newRelObjs = append(newRelObjs, updatedRelatedObjs...)
+
+		// sort the related objects by kind and name
+		sort.SliceStable(newRelObjs, func(i, j int) bool {
+			if newRelObjs[i].Object.Kind != newRelObjs[j].Object.Kind {
+				return newRelObjs[i].Object.Kind < newRelObjs[j].Object.Kind
+			}
+
+			return newRelObjs[i].Object.Metadata.Name < newRelObjs[j].Object.Metadata.Name
+		})
+
+		policy.Status.RelatedObjects = newRelObjs
+	}
+
+	if condChanged || relObjsChanged {
+		return r.Status().Update(ctx, policy)
+	}
+
+	return nil
+}
+
+func conditionChanged(updatedCondition, existingCondition metav1.Condition) bool {
+	if updatedCondition.Message != existingCondition.Message {
+		return true
+	}
+
+	if updatedCondition.Reason != existingCondition.Reason {
+		return true
+	}
+
+	if updatedCondition.Status != existingCondition.Status {
+		return true
+	}
+
+	return false
+}
+
+// The Compliance condition is calculated by going through the known conditions in a consistent
+// order, checking if there are any reasons the policy should be NonCompliant, and accumulating
+// the reasons into one string to reflect the whole status.
+func calculateComplianceCondition(policy *policyv1beta1.OperatorPolicy) metav1.Condition {
+	foundNonCompliant := false
+	messages := make([]string, 0)
+
+	idx, cond := policy.Status.GetCondition(opGroupConditionType)
+	if idx == -1 {
+		messages = append(messages, "the status of the OperatorGroup is unknown")
+		foundNonCompliant = true
+	} else {
+		messages = append(messages, cond.Message)
+
+		if cond.Status != metav1.ConditionTrue {
+			foundNonCompliant = true
+		}
+	}
+
+	idx, cond = policy.Status.GetCondition(subConditionType)
+	if idx == -1 {
+		messages = append(messages, "the status of the Subscription is unknown")
+		foundNonCompliant = true
+	} else {
+		messages = append(messages, cond.Message)
+
+		if cond.Status != metav1.ConditionTrue {
+			foundNonCompliant = true
+		}
+	}
+
+	// FUTURE: check additional conditions
+
+	if foundNonCompliant {
+		return metav1.Condition{
+			Type:               compliantConditionType,
+			Status:             metav1.ConditionFalse,
+			LastTransitionTime: metav1.Now(),
+			Reason:             "NonCompliant",
+			Message:            "NonCompliant; " + strings.Join(messages, ", "),
```

Comment:
Optional: would a semicolon as the join separator be better?

Reply:
I think the comma more closely matches what config-policy does.